### PR TITLE
feat: statsig async initialization [OTE-496]

### PR DESCRIPTION
**Pull request by yogurtandjam**
Fixes the issue that was called out in https://github.com/dydxprotocol/v4-web/pull/756

The race condition was not inherent to abacus. 
The condition was that the client is constructed separately from initialization, so we were early returning an uninitalized client. This caused abacus to have split behavior where it would call the squid assets endpoint but use the skip processors.

We fix this problem by instead returning the initialization promise instead

TLDR; I am out of practice with promises. This should not have been that hard to debug 😵 

PS: TODO - think about adding tests for this to make sure we don't break the async functionality in the future

### CODE REVIEW

File: src/lib/statsig.ts [hunk 1]
```diff
@@ -1,20 +1,48 @@
...
-// https://linear.app/dydx/project/feature-experimentation-6853beb333d7/overview
+/**
+ *
+ * This is used only in useInitializePage to block abacus start on the fully loaded async statsig config.
+ * This prevents new users from automatically received a 'false' for all feature gates because they
+ * don't have a statsig value cached.
+ */
+export const statsigConfigPromise = async () => {
+  const client = await initStatsig();
+  const allGateValues = Object.values(StatSigFlags).reduce((acc, gate) => {
```
Comment (tyleroooo) [inline src/lib/statsig.ts:44]: Cleaner: 
`
const allGateValues = mapValues(StatSigFlags, (gate) => client.checkGate(gate));
`

Comment (tyleroooo) [reply src/lib/statsig.ts:44]: (if I am understanding correctly)

File: src/hooks/useStatsig.tsx [hunk 2]
```diff
@@ -7,22 +7,18 @@ import {
...
 export const StatsigProvider = ({ children }: { children: React.ReactNode }) => {
+  const [statsigClient, setStatsigClient] = useState<StatsigClient | null>(null);
+  useEffect(() => {
+    const init = async () => {
+      const client = await initStatsig();
+      setStatsigClient(client);
+    };
+    init();
+  }, []);
+  if (!statsigClient) return <div>{children}</div>;
```
Comment (tyleroooo) [inline src/hooks/useStatsig.tsx:21]: Will consumers fail gracefully if they assume there's provider data on the context and it's not there?

File: src/lib/statsig.ts (hunk 1 above)
Comment (yogurtandjam) [reply src/lib/statsig.ts:44]: we're reducing over the `StatsigFlags` values, so i think `mapValues` won't give us what we want, because it will be mapping over the keys of `StatsigFlags` (if i'm understanding the lodash docs correctly)

File: src/hooks/useStatsig.tsx (hunk 2 above)
Comment (yogurtandjam) [reply src/hooks/useStatsig.tsx:21]: yup, for example i can change the provider to never actually return the StatsingProvider
```
if (!statsigClient) return <div>{children}</div>;
  return <div>{children}</div>;
  // return <StatsigProviderInternal client={statsigClient}> {children} </StatsigProviderInternal>;
```

and yet on the App.tsx i can call
```
const x = useStatsigGateValue(StatSigFlags.ffSkipMigration);
console.log('hey', x);
```

which console logs
Screenshot 2024-07-03 at 12 33 02 PM


i think the one thing i'm not sure about is if the statsig client fails to initialize. this should never happen though since the initialization happens locally even if the async call fails (i.e. even when i turn off my wifi i can still load v4-web just fine)

File: src/lib/statsig.ts [hunk 3]
```diff
@@ -1,20 +1,69 @@
...
+ * This is used only in useInitializePage to block abacus start on the fully loaded async statsig config.
+ * This prevents new users from automatically received a 'false' for all feature gates because they
+ * don't have a statsig value cached.
+ *
+ * Try catches in case the statsig init or gatechecks fail.
+ */
+export const statsigConfigPromise = async (): Promise<StatsigConfigType> => {
+  try {
+    const client = await initStatsig();
+    const [firstGateResults, ...otherGateResults] = Object.values(StatSigFlags).map((gateId) => ({
```
Comment (yogurtandjam) [inline src/lib/statsig.ts:60]: @tyleroooo invert messes up the typing so it's hard to be certain we're using the correct gateId's so i instead used a map + merge. LMK if this works for you :)

File: src/lib/statsig.ts [hunk 4]
```diff
@@ -1,20 +1,69 @@
...
+/**
+ * This method creates a promise and assigns it to the variable initPromise.
+ * If an initPromise has already been generated, it returns the existing promise.
+ * This causes all calls to this function to always return the statsigClient only after
+ * it has retrieved all values from statsig
+ *
+ * @returns initPromise: a promise that returns the statsig client only after it is initialized
+ */
+export const initStatsig = async () => {
+  if (initPromise) {
```
[the remark this answers is not in the document]
Comment (yogurtandjam) [reply src/lib/statsig.ts:19]: yeah it's a little bit annoying. i spent some time thinking about this last night but the only other thing i coudl think of would be to put the entire app into a loading state while the promise was running, which i think is a cleaner but possibly bit trickier solution since we'd also need to put prevent all the providers from running as well 🤔.

FWIW i have a project to bolster our experimentation https://linear.app/dydx/project/feature-experimentation-6853beb333d7/issues so illl be investigating this later.

File: src/lib/statsig.ts [hunk 5]
```diff
@@ -1,20 +1,69 @@
...
+
+/**
+ *
+ * This is used only in useInitializePage to block abacus start on the fully loaded async statsig config.
+ * This prevents new users from automatically received a 'false' for all feature gates because they
+ * don't have a statsig value cached.
+ *
+ * Try catches in case the statsig init or gatechecks fail.
+ */
+export const statsigConfigPromise = async (): Promise<StatsigConfigType> => {
```
Comment (tyleroooo) [inline src/lib/statsig.ts:57]: let's not call this promise since it's not a promise, it's an async function. Can call it initStatsigAsync or something.

File: src/lib/statsig.ts (hunk 3 above)
Comment (tyleroooo) [reply src/lib/statsig.ts:60]: what about object.assign({}, allObjects)

Comment (tyleroooo) [reply src/lib/statsig.ts:60]: or merge ({}, allObjects)

File: src/lib/statsig.ts (hunk 5 above)
Comment (yogurtandjam) [reply src/lib/statsig.ts:57]: oh good point, this is a weird name

**Review by tyleroooo — APPROVED** [review]
didn't test but code seems good

**Review by rosepuppy — APPROVED** [review]
